File: rti_connext_dds-5.2.3/resource/template/rti_workspace/examples/connext_dds/c++/hello_world_xml_compiled/HelloWorld_publisher.cxx
/* HelloWorld_publisher.cxx

   A publication of data of type HelloWorld   

       
modification history
------------ -------       
*/

#include <stdio.h>
#include <stdlib.h>
#ifdef RTI_VX653
#include <vThreadsData.h>
#endif

#include "ndds/ndds_cpp.h"
#include "HelloWorldSupport.h"

/* Delete all entities */
int publisherShutdown(
    DDSDomainParticipant * participant) {
    DDS_ReturnCode_t retCode;
    int status = 0;

    if (participant != NULL) {
        retCode = participant->delete_contained_entities();
        if (retCode != DDS_RETCODE_OK) {
            printf("delete_contained_entities error %d\n", retCode);
            status = -1;
        }

        retCode = DDSTheParticipantFactory->delete_participant(participant);
        if (retCode != DDS_RETCODE_OK) {
            printf("delete_participant error %d\n", retCode);
            status = -1;
        }
    }

    /* RTI Connext provides finalize_instance() method on
       domain participant factory for people who want to release memory used
       by the participant factory. Uncomment the following block of code for
       clean destruction of the singleton. */
/*
    retcode = DDSDomainParticipantFactory::finalize_instance();
    if (retcode != DDS_RETCODE_OK) {
        printf("finalize_instance error %d\n", retcode);
        status = -1;
    }
*/

    return status;
}

int publisherMain(int sampleCount) {
    DDS_ReturnCode_t retCode;
    int count = 0;  
    DDS_Duration_t sendPeriod = {4,0};

    /* type registration */
    retCode = DDSTheParticipantFactory->register_type_support(
                                        HelloWorldTypeSupport::register_type, 
                                        "HelloWorldType");
    if (retCode != DDS_RETCODE_OK) {
        printf("register_type_support %d\n", retCode);
        publisherShutdown(NULL);
        return -1;
    }
    
    /* To customize participant QoS, use 
       the configuration file USER_QOS_PROFILES.xml */   
    DDSDomainParticipant * participant = DDSTheParticipantFactory->
            create_participant_from_config("MyParticipantLibrary::PublicationParticipant");
    if (participant == NULL) {
        printf("create_participant_from_config error\n");
        publisherShutdown(participant);
        return -1;
    }
    
    HelloWorldDataWriter * helloWorldWriter = HelloWorldDataWriter::narrow(
            participant->lookup_datawriter_by_name(
                                      "MyPublisher::HelloWorldWriter"));    
    if (helloWorldWriter == NULL) {
        printf("lookup_datawriter_by_name error\n");
        publisherShutdown(participant);
        return -1;
    }
    
    HelloWorld * helloWorldData = HelloWorldTypeSupport::create_data();   
    if (helloWorldData == NULL) {
        printf("HelloWorldTypeSupport_create_data error\n");
        publisherShutdown(participant);
        return -1;
    }
            
    /* Main loop */
    for (count=0; (sampleCount == 0) || (count < sampleCount); ++count) {
        printf("Writing HelloWorld, count: %d\n", count);

        /* Set the data fields */
		strncpy(helloWorldData->sender,  "John Smith", MAX_NAME_LEN); 
		strncpy(helloWorldData->message, "Hello World!", MAX_MSG_LEN); 
		helloWorldData->count   = count; 

        retCode = helloWorldWriter->write(*helloWorldData, DDS_HANDLE_NIL);
        if (retCode != DDS_RETCODE_OK) {
            printf("write error %d\n", retCode);
            publisherShutdown(participant);
            return -1;
        }
        
        NDDSUtility::sleep(sendPeriod);
    }

    /* Delete data sample */
    HelloWorldTypeSupport::delete_data(helloWorldData);

    /* Delete all entities */
    return publisherShutdown(participant);
}
 
int main(int argc, char *argv[])
{
    int sampleCount = 0; /* infinite loop */   
    
    if (argc >= 2) {
        sampleCount = atoi(argv[1]);
    }

    /* Uncomment this to turn on additional logging
    NDDSConfigLogger::get_instance()->
        set_verbosity_by_category(NDDS_CONFIG_LOG_CATEGORY_API, 
                                  NDDS_CONFIG_LOG_VERBOSITY_STATUS_ALL);
    */
    
    return publisherMain(sampleCount);
}

File: ros2_ws/rti_workspace/5.2.3/examples/connext_dds/cs/hello_world_xml_compiled/HelloWorld_publisher.cs
/* ****************************************************************************
 *         (c) Copyright, Real-Time Innovations, All rights reserved.       
 *                                                                          
 *         Permission to modify and use for internal purposes granted.      
 * This software is provided "as is", without warranty, express or implied. 
 *                                                                          
 * ****************************************************************************
 */

using System;
using System.Collections.Generic;
using System.Text;

namespace Hello
{
    /**
     * Manage the data publication for the Hello example 
     */
    class HelloPublisher
    {        
        DDS.DomainParticipant participant;
        HelloWorldDataWriter helloWriter;
        HelloWorld instance;

        public void Dispose()
        {
            
            if (this.participant != null)
            {
                this.participant.delete_contained_entities();
                this.helloWriter = null;
                DDS.DomainParticipantFactory.get_instance().delete_participant(
                    ref this.participant);                
            }

            if (instance != null) {
                HelloWorldTypeSupport.delete_data(instance);
                instance = null;
            }

            DDS.DomainParticipantFactory.finalize_instance();
        }

        public void publisherStart(
            int sampleCount) 
        { 
                       
            /* To customize QoS, use 
            the configuration file USER_QOS_PROFILES.xml */
            if (this.participant == null)
            {
                DDS.DomainParticipantFactory.get_instance().register_type_support(
                    HelloWorldTypeSupport.get_instance(),
                    "HelloWorldType");
                this.participant = DDS.DomainParticipantFactory.get_instance().
                        create_participant_from_config(
                                        "MyParticipantLibrary::PublicationParticipant");
                if (this.participant == null)
                {
                    Console.Error.WriteLine("! Unable to create DDS domain participant");
                    return;
                }
            }

            if (this.helloWriter == null)
            {
                this.helloWriter = (HelloWorldDataWriter)
                        this.participant.lookup_datawriter_by_name(
                                                  "MyPublisher::HelloWorldWriter");
                if (this.helloWriter == null)
                {
                    Console.Error.WriteLine("! Unable to get DDS HelloWorld writer");
                    return;
                }
            }

            if (this.instance == null)
            {
                this.instance = HelloWorldTypeSupport.create_data();
                if (this.instance == null)
                {
                    Console.Error.WriteLine("! Unable to create DDS HelloWorld instance");
                    return;
                } 
            }              
       
            /* Main loop */
            for (int count=0; (sampleCount == 0) || (count < sampleCount); ++count) 
            {
                Console.WriteLine("Writing HelloWorld, count: " + count);
                this.instance.sender = "John Smith";
                this.instance.message = "Hello World!";
                this.instance.count = count;
               
                try
                {
                    this.helloWriter.write(this.instance, ref DDS.InstanceHandle_t.HANDLE_NIL);
                }
                catch (DDS.Retcode_Error e)
                {
                    Console.Error.WriteLine("! Write error " +
                                e.GetType().ToString() + ": " + e.Message);
                    return;
                }

                System.Threading.Thread.Sleep(3000);
            }
        }

        public static void Main(string[] argv) 
        {

            int sampleCount = 0; // infinite loop

            if (argv.Length >= 1)
            {
                sampleCount = Int32.Parse(argv[0]);
            }

            HelloPublisher publisher = new HelloPublisher();            
            publisher.publisherStart(sampleCount);
            publisher.Dispose();
        }        
        
    }
}
